# Groundwater pH and bicarbonate

**Instructor solution.** Run this notebook independently from top to bottom.

In [1]:
from pathlib import Path
import sys

# Support kernels started in the repository root or in instructor-notebooks.
ROOT = Path.cwd()
if not (ROOT / "data").is_dir():
    ROOT = ROOT.parent
if not (ROOT / "data").is_dir():
    raise FileNotFoundError("Open this notebook from the repository or instructor-notebooks folder.")
sys.path.insert(0, str(ROOT))

In [2]:
import pandas as pd
import statsmodels.formula.api as smf
df = pd.read_csv(ROOT / "data/ground_water.csv")
df.head()

,pH,Bicarb
0,7.6,157
1,7.1,174
2,8.2,175
3,7.5,188
4,7.4,171


In [3]:
regression = smf.ols("pH ~ Bicarb", data=df).fit()
regression.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     pH   R-squared:                       0.115
Model:                            OLS   Adj. R-squared:                  0.088
Method:                 Least Squares   F-statistic:                     4.169
Date:                Tue, 29 Sep 2026   Prob (F-statistic):             0.0495
Time:                        21:12:35   Log-Likelihood:                -22.191
No. Observations:                  34   AIC:                             48.38
Df Residuals:                      32   BIC:                             51.43
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      8.0976      0.229     35.405      0.000       7.632       8.563
Bicarb        -0.0031      0.001     -2.042      0.049      -0.006   -7.34e-06
==============================================================================
Omnibus:                        0.527   Durbin-Watson:                   2.377
Prob(Omnibus):                  0.768   Jarque-Bera (JB):                0.423
Skew:                           0.256   Prob(JB):                        0.809
Kurtosis:                       2.810   Cond. No.                         426.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [4]:
print(regression.params)
print("R-squared:", regression.rsquared)
print("Observations used:", int(regression.nobs))

Intercept    8.097595
Bicarb      -0.003052
dtype: float64
R-squared: 0.11526739372275296
Observations used: 34


The fitted equation is pH ≈ 8.0976 − 0.003052 × Bicarb (ppm). A 100 ppm increase in bicarbonate corresponds to a fitted decrease of about 0.305 pH units. R² ≈ 0.115 means the model explains about 11.5% of the observed pH variation. This association alone does not establish causation, and bicarbonate alone leaves most variation unexplained. Retain an intercept: pH need not be zero when bicarbonate is zero.